text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,513 @@
+
+### Client Modes
+
+OTLP defines 2 modes of operation for clients: Unary and Streaming.
+
+Unary mode is simpler to implement and provides good performance in sub-millisecond latency networks, at the cost of worse throughput in high latency networks.
+
+Streaming mode requires somewhat more complicated implementation but provides significantly better throughput on high latency networks. 
+
+Clients are recommended to implemented both modes but may choose to implement only one.
+
+Servers MUST implement both modes. Clients operating in different modes can connect to the same server.
+
+#### Unary
+
+In Unary mode after sending the request the client MUST wait until the response is received from the server. There can be at most only one request in flight that is not yet acknowledged by the server. The server MUST reply with `Export` response that includes the ID of the last request. Until such response or an error from underlying transport layer is received the client MUST NOT send any other requests. Unary mode is implemented using gRPC unary RPC.
+
+![Unary](images/otlp-unary.png)
+
+Unary mode is recommended when simplicity of implementation is desirable and when the client and the server are connected via very low-latency network, such as for example when the client is an instrumented application and the server is a OpenTelemetry Service running as a local daemon.
+
+The maximum achievable throughput in unary mode is `max_request_size / (network_latency + server_response_time)`. For example if the request can contain at most 100 spans, network roundtrip latency is 200ms and server response time is 300 ms, then the maximum achievable throughput is `100 spans / (200ms+300ms)` or 200 spans per second. It is easy to see that in high latency networks or when the server response time is high it is difficult to achieve good throughput unless requests are very big (even when client, server and network throughput potentially allow much bigger processing rate).
+
+The ID field of the Export request and response is not used in Unary mode since the underlying transport already supports acknowledgements and one-to-one mapping of requests and responses.
+
+#### Streaming
+
+In Streaming mode the client may send multiple `Export` requests without waiting for the response to the earlier sent requests, essentially creating a pipeline of requests that are currently in flight that are not acknowledged.
+
+`Export` request includes an ID field that is unique for each request. The value of ID field is a sequential uint64 number generated by the client. The server MUST respond and acknowledge each `Export` request with a corresponding response that includes the ID of the request. Note that the order of IDs in a sequence of responses coming from the server may be different than the order of IDs in a sequence of earlier sent requests coming from the client. The client MUST be able to correctly process such out-of-order responses.
+
```

Comment:
Why do you use generated IDs and support out-of-order acking like this?

Why not use a design that's more similar to how TCP acks with the sequence number of the last byte received and have the server send acks with the number of the last span it read from the stream? That would
* Not impose artificial batching and allow for allow for clients to open multiple streams and redirect spans/metrics at a granular level if it exceeds the throughput of a single stream.
* Simplify the client code since there's no need to maintain an indexed structure of the batches sent.
* Reduce the amount of data exchange (would require fewer ack responses from the server).
Is there a use case for out-of-order acking?

Reply:
@rnburn you are right that we could do acks similar to TCP. However acks are not the only reason that we want to send responses. Responses also carry the result code and we need one result code per request sent. We either have to send a response for each request or batch responses and send at once but that both complicates the implementation and requires the client wait for a batched response. This is unneccessary and it is much simpler to send one response per request.

To address your specific concerns:
>Not impose artificial batching

I am not sure why you call batching "artificial". There are several reasons why batching can be a good idea. Batching for the purpose of reducing number of requests is only one of them (another benefit is for example better compression ratio because larger data usually compresses more).

> and allow for allow for clients to open multiple streams and redirect spans/metrics at a granular level if it exceeds the throughput of a single stream.

Nothing in this proposal prevents implementations to open multiple streams if needed. In fact my reference implementation does that during stream re-openinig transition period. Implementations can easily use multiple streams using described protocol.

>Simplify the client code since there's no need to maintain an indexed structure of the batches sent.

True, however for clients where simplicity is important there is Unary mode. Streaming mode's primary focus is not implementation simplicity.

>Reduce the amount of data exchange (would require fewer ack responses from the server).

For typical use cases this is negligible. Telemetry data size is typically much larger than the tiny response.

Another benefit of sending acks immediately is that it allows the client to earlier free request data that is kept in-memory.

>Is there a use case for out-of-order acking?

Yes, there is. For example server implementation can be done using thread pools (task pools) that consume received requests, process concurrently and send a response back. This normally means the order of request processing _completion_ is undefined. To use sequential ack approach that you suggest one needs to perform ordering of completions or corresponding responses which is additional complication.

In general what you suggest is probably another approach that could be researched, and it will likely result in different tradeoffs (more complex server, more simple client). I beleive my proposal hits the sweet spot that I wanted to hit based on my previous experience with OpenCensus.

Reply:
Why do we need a concept of requests for streaming at all? 

Here's how I'd see a streaming protocol working: 
1. As spans/metrics are finished, they're serialized, framed, and added into a buffer.
2. When a reporter is ready to sent data to a collection endpoint, it uses writev with the already framed and serialized spans to write the data directly to the collection endpoint's socket.

There's no need to bundle spans into requests and the recorder is low overhead because there's no serialization cost. If you wanted to add acking to this approach, you could have the collection endpoint regularly send back ack responses, at points of its choosing, denoting how far in the stream it's processed data.

> Nothing in this proposal prevents implementations to open multiple streams if needed. In fact my reference implementation does that during stream re-openinig transition period. Implementations can easily use multiple streams using described protocol.

I don't claim that it can't use multiple streams; however, you lose the ability to granularly redirect data. 

Suppose I have multiple span streams, a buffer of 1k spans, and I write the data to a non-blocking socket. If the non-blocking call only wrote 200 spans, then I can immediately start writing the remaining full spans to other streams. If, instead, I have to write batches of spans, then I lose this ability to pick up writing where last write call finished. 

Being able to granularly load balance like this allows throughput of the streams to naturally adjust to the rate that the endpoint is processing the data.

Furthermore, if all the sockets are blocked, I can block with epoll until one of the stream's sockets is writable again, and pick up writing with not just the remaining data, but also include any additional data that's been added into to the buffer between the last write call -- there's no need to batch the new data first.

> In general what you suggest is probably another approach that could be researched, and it will likely result in different tradeoffs (more complex server, more simple client). I beleive my proposal hits the sweet spot that I wanted to hit based on my previous experience with OpenCensus.

Maybe there are some cases where batched request/response streaming makes sense. But IMO eagerly serializing (i.e. writing the span/metrics serialization as methods on it are called) combined with streaming data unbatched over a raw socket is a definite sweet spot that leads to performant, high throughput tracers, and it should be supported.

Reply:
@rnburn 
>Why do we need a concept of requests for streaming at all?

That's how gRPC conceptually works and I suggest to use gRPC as a transport.

Other transports are certainly possible. My selection criteria was ease of implementation, ubiquity of client and server libraries, proven robustness, extensibility and out-of-the-box availability of necessary functionality and gRPC ticked the boxes.

If you would like to propose a different transport and a different protocol feel free to submit an alternate RFC. FYI, these are the design goals and requirements for a general-purpose telemetry data protocol that have been previously discussed and agreed on:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/proto-design-goals.md
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/proto-requirements.md

I know you earlier suggested https://github.com/open-telemetry/opentelemetry-cpp/issues/2 and that looks interesting to me as a specialized approach but I believe it doesn't satisfy some of the requirements for general-purpose protocol. If you would like to work further on that and propose as a protocol I'll be glad to review and comment.